src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/EnumConverter.cs

```diff
@@ -381,25 +311,42 @@ internal override void WriteAsPropertyNameCore(Utf8JsonWriter writer, T value, J
                 : (rentedBuffer = ArrayPool<char>.Shared.Rent(bufferLength));
 
             int charsWritten = reader.CopyString(charBuffer);
-            ReadOnlySpan<char> source = charBuffer.Slice(0, charsWritten);
+            Span<char> source = charBuffer.Slice(0, charsWritten);
+#else
+            string source = reader.GetString();
 #endif
+            // Skip the built-in enum parser and go directly to the read cache if either:
+            //
+            // 1. one of the enum fields have had their names overridden OR
+            // 2. the source string represents a number when numbers are not permitted.
+            //
+            // For backward compatibility reasons the built-in parser is not skipped if a naming policy is specified.
+            bool skipEnumParser = _containsNameAttributes ||
+                ((_converterOptions & EnumConverterOptions.AllowNumbers) is 0 && JsonHelpers.IntegerRegex.IsMatch(source));
 
-            bool success;
-            if ((_converterOptions & EnumConverterOptions.AllowNumbers) != 0 || !JsonHelpers.IntegerRegex.IsMatch(source))
+            if (!skipEnumParser && Enum.TryParse(source, ignoreCase: true, out value))
             {
-                // Try parsing case sensitive first
-                success = Enum.TryParse(source, out value) || Enum.TryParse(source, ignoreCase: true, out value);
+                success = true;
+                goto End;
             }
-            else
+
+            Debug.Assert(_nameCacheForReading is null == (_namingPolicy is null && !_containsNameAttributes),
+                         "A read cache should only be populated if we have a naming policy or name attributes.");
+
+            if (_nameCacheForReading is not { } nameCacheForReading)
```

Comment:
```suggestion
            if (_nameCacheForReading is null)
```

Reply:
It feels weird that a cache decides whether something fails to parse.

Reply:
I could have replaced it with the RHS of the assertion in line 333 which is equivalent. But ultimately it's more instructions compared to checking if the cache is null.